pkg/dtest/k3s.go

```diff
@@ -0,0 +1,312 @@
+	"componentstatuses",
+	"configmaps",
+	"endpoints",
+	"events",
+	"limitranges",
+	"namespaces",
+	"nodes",
+	"persistentvolumeclaims",
+	"persistentvolumes",
+	"pods",
+	"podtemplates",
+	"replicationcontrollers",
+	"resourcequotas",
+	"secrets",
+	"serviceaccounts",
+	"services",
+	"mutatingwebhookconfigurations.admissionregistration.k8s.io",
+	"validatingwebhookconfigurations.admissionregistration.k8s.io",
+	"customresourcedefinitions.apiextensions.k8s.io",
+	"apiservices.apiregistration.k8s.io",
+	"controllerrevisions.apps",
+	"daemonsets.apps",
+	"deployments.apps",
+	"replicasets.apps",
+	"statefulsets.apps",
+	"tokenreviews.authentication.k8s.io",
+	"localsubjectaccessreviews.authorization.k8s.io",
+	"selfsubjectaccessreviews.authorization.k8s.io",
+	"selfsubjectrulesreviews.authorization.k8s.io",
+	"subjectaccessreviews.authorization.k8s.io",
+	"horizontalpodautoscalers.autoscaling",
+	"cronjobs.batch",
+	"jobs.batch",
+	"certificatesigningrequests.certificates.k8s.io",
+	"leases.coordination.k8s.io",
+	"daemonsets.extensions",
+	"deployments.extensions",
+	"ingresses.extensions",
+	"networkpolicies.extensions",
+	"podsecuritypolicies.extensions",
+	"replicasets.extensions",
+	"helmcharts.helm.cattle.io",
+	"addons.k3s.cattle.io",
+	"listenerconfigs.k3s.cattle.io",
+	"ingresses.networking.k8s.io",
+	"networkpolicies.networking.k8s.io",
+	"runtimeclasses.node.k8s.io",
+	"poddisruptionbudgets.policy",
+	"podsecuritypolicies.policy",
+	"clusterrolebindings.rbac.authorization.k8s.io",
+	"clusterroles.rbac.authorization.k8s.io",
+	"rolebindings.rbac.authorization.k8s.io",
+	"roles.rbac.authorization.k8s.io",
+	"priorityclasses.scheduling.k8s.io",
+	"csidrivers.storage.k8s.io",
+	"csinodes.storage.k8s.io",
+	"storageclasses.storage.k8s.io",
+	"volumeattachments.storage.k8s.io",
+}
+
+func isK3sReady() bool {
+	kubeconfig := getKubeconfigPath()
+
+	if kubeconfig == "" {
+		return false
+	}
+
+	cmd := supervisor.Command(prefix, "kubectl", "--kubeconfig", kubeconfig, "api-resources", "-o", "name")
+	output, err := cmd.Capture(nil)
+	if err != nil {
+		return false
+	}
+	resources := make(map[string]bool)
+	for _, line := range strings.Split(output, "\n") {
+		resources[strings.TrimSpace(line)] = true
+	}
+
+	for _, req := range requiredResources {
+		_, exists := resources[req]
+		if !exists {
+			return false
+		}
+	}
+
+	return true
+}
+
+const k3sConfigPath = "/etc/rancher/k3s/k3s.yaml"
+
+// GetKubeconfig returns the kubeconfig contents for the running k3s
+// cluster as a string. It will return the empty string if no cluster
+// is running.
+func GetKubeconfig() string {
+	if !isKubeconfigReady() {
+		return ""
+	}
+
+	id := tag2id("k3s")
+
+	if id == "" {
+		return ""
+	}
+
+	cmd := supervisor.Command(prefix, "sh", "-c", fmt.Sprintf("docker cp \"%s:%s\" - | tar -xO", id, k3sConfigPath))
+	kubeconfig := cmd.MustCapture(nil)
+	kubeconfig = strings.ReplaceAll(kubeconfig, "localhost:6443", fmt.Sprintf("%s:%s", dockerIP(), k3sPort))
```

Comment:
I wrote yesterday:
> Use `net.JoinHostPort(dockerIP(), k3sPort)` instead of `fmt.Sprintf("%s:%s", dockerIP(), k3sPort)`.  If `dockerIP()` is ever modified to return an IPv6 address, the Sprintf solution won't behave correctly.

Reply:
Done